src/v/config/property_constraints.cc

```diff
@@ -0,0 +1,462 @@
+  ss::sstring& name,
+  std::optional<ss::sstring> min_str_opt,
+  std::optional<ss::sstring> max_str_opt,
+  constraint_enabled_t enabled) {
+    auto& cfg = config::shard_local_cfg();
+
+    if (name == "log_cleanup_policy") {
+        return std::make_shared<
+          constraint_validator_enabled<model::cleanup_policy_bitflags>>(
+          std::move(enabled),
+          cfg.log_cleanup_policy.bind(),
+          [](const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.cleanup_policy_bitflags
+                       ? *topic_cfg.properties.cleanup_policy_bitflags
+                       : model::cleanup_policy_bitflags::none;
+          },
+          [](
+            cluster::topic_configuration& topic_cfg,
+            model::cleanup_policy_bitflags& topic_val) {
+              topic_cfg.properties.cleanup_policy_bitflags = topic_val;
+          });
+    } else if (name == "log_retention_ms") {
+        std::optional<std::chrono::milliseconds> min_opt{std::nullopt},
+          max_opt{std::nullopt};
+        if (min_str_opt) {
+            min_opt = std::chrono::milliseconds{std::stol(*min_str_opt)};
+        }
+        if (max_str_opt) {
+            max_opt = std::chrono::milliseconds{std::stol(*max_str_opt)};
+        }
+        auto def_val = *cfg.log_retention_ms.default_value();
+        return std::make_shared<
+          constraint_validator_range<std::chrono::milliseconds>>(
+          std::move(min_opt),
+          std::move(max_opt),
+          [def_val{std::move(def_val)}](
+            const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.retention_duration
+                         .has_optional_value()
+                       ? topic_cfg.properties.retention_duration.value()
+                       : def_val;
+          },
+          [](
+            cluster::topic_configuration& topic_cfg,
+            std::chrono::milliseconds& topic_val) {
+              topic_cfg.properties.retention_duration
+                = tristate<std::chrono::milliseconds>(topic_val);
+          });
+    } else if (name == "log_segment_ms") {
+        std::optional<std::chrono::milliseconds> min_opt{std::nullopt},
+          max_opt{std::nullopt};
+        if (min_str_opt) {
+            min_opt = std::chrono::milliseconds{std::stol(*min_str_opt)};
+        }
+        if (max_str_opt) {
+            max_opt = std::chrono::milliseconds{std::stol(*max_str_opt)};
+        }
+        auto def_val = *cfg.log_segment_ms.default_value();
+        return std::make_shared<
+          constraint_validator_range<std::chrono::milliseconds>>(
+          std::move(min_opt),
+          std::move(max_opt),
+          [def_val{std::move(def_val)}](
+            const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.segment_ms.has_optional_value()
+                       ? topic_cfg.properties.segment_ms.value()
+                       : def_val;
```

Comment:
issue: `has_optional_value()` returns true if `segment_ms` holds a `std::optional<>`.  Call to `value()` returns `std::optional::value` wihtout any checks and could lead to an error.

Reply:
According to tristate.h, `has_optional_value()` returns true when the tristate is not disabled and the optional holds a value. https://github.com/redpanda-data/redpanda/blob/db4e7f33e41c66cea98fc4541b2caa628d59f25d/src/v/tristate.h#L51

This is also how the output operator checks for value in https://github.com/redpanda-data/redpanda/blob/db4e7f33e41c66cea98fc4541b2caa628d59f25d/src/v/tristate.h#L106

So I am not sure I understand the issue here.

Reply:
I must have misread that.  My bad